src/Tools/dotnet-monitor/Egress/EgressService.cs

```diff
@@ -16,31 +21,129 @@ namespace Microsoft.Diagnostics.Tools.Monitor.Egress
+                throw new EgressException(string.Format(CultureInfo.CurrentCulture, Strings.ErrorMessage_EgressProviderTypeNotRegistered, providerName));
             }
-            throw new EgressException(string.Format(CultureInfo.CurrentCulture, Strings.ErrorMessage_EgressProviderDoesNotExist, providerName));
+
+            // Get the egress provider that matches the options type and return the weaker-typed
+            // interface in order to allow egressing from the service without having to use reflection
+            // to invoke it.
+            return (IEgressProviderInternal)_serviceProvider.GetRequiredService(
+                typeof(IEgressProviderInternal<>).MakeGenericType(optionsType));
         }
 
-        public Task<EgressResult> EgressAsync(string providerName, Func<Stream, CancellationToken, Task> action, string fileName, string contentType, IEndpointInfo source, CancellationToken token)
+        private static EgressArtifactSettings CreateSettings(IEndpointInfo source, string fileName, string contentType)
         {
-            if (_egressOptions.CurrentValue.Providers.TryGetValue(providerName, out ConfiguredEgressProvider provider))
+            EgressArtifactSettings settings = new();
+            settings.Name = fileName;
+            settings.ContentType = contentType;
+
+            // Activity metadata
+            Activity activity = Activity.Current;
+            if (null != activity)
             {
-                return provider.EgressAsync(action, fileName, contentType, source, token);
+                settings.Metadata.Add(
+                    ActivityMetadataNames.ParentId,
+                    activity.GetParentId());
+                settings.Metadata.Add(
+                    ActivityMetadataNames.SpanId,
+                    activity.GetSpanId());
+                settings.Metadata.Add(
+                    ActivityMetadataNames.TraceId,
+                    activity.GetTraceId());
+            }
+
+            // Artifact metadata
+            settings.Metadata.Add(
+                ArtifactMetadataNames.ArtifactSource.ProcessId,
+                source.ProcessId.ToString(CultureInfo.InvariantCulture));
+            settings.Metadata.Add(
+                ArtifactMetadataNames.ArtifactSource.RuntimeInstanceCookie,
+                source.RuntimeInstanceCookie.ToString("N"));
+
+            return settings;
+        }
+
+        private void Reload()
+        {
+            _optionsTypeMap.Clear();
+            _providerTypeMap.Clear();
+            foreach (IEgressProviderConfigurationProvider provider in _providers)
+            {
+                _optionsTypeMap.Add(provider.ProviderType, provider.OptionsType);
+
+                foreach (IConfigurationSection optionsSection in provider.Configuration.GetChildren())
+                {
+                    _providerTypeMap.Add(optionsSection.Key, provider.ProviderType);
```

Comment:
This will throw if two providers have the same name e.g.:

```json
"Egress": {
    "AzureBlobStorage": {
        "SameName": { }
    },
    "FileSystem": {
        "SameName": {},
    },
}
```

Probably should use `TryAdd` and then log when a duplicate is specified.

Reply:
Fixed. It keeps the first one declared, drops subsequent ones, and logs a message describing precisely what action is taken.